### ISSUE TITLE
SoC sequence looks suspicious - reserve energy not actually precharged

### ISSUE THREAD

**Issue by EVGrokker** [aesculus/EVTO-App-Feedback#429]
>Car Data: X90D, Charger: 48, Wheels: 20, Tires: 0, Pano: 0, 5 Seat: 1, Rear CC: 0, Bat Life: 94 
Device Info: iPhone; CPU iPhone OS 10_3_2 like Mac OS X 
Settings: Reserve SoC:20, Default SoC: 90, Region: 0, Units: 0 
Version: 1.2.0 (35) 
IAP: Pro:1, Sub:1 
Location: Unknown 
Trip: 21673 Paloma Drive, [t76d93v2iqg0puzqa3c6k5]

Take a look at the last charging stop in Bend, the destination arrival SoC, and the reserve energy. It doesn't make sense that you could leave the Bend supercharger with 33% (according to the Waypoint Inspector) and arrive at the nearby final destination with 32% SoC (that's plausible) *and* a 19% reserve allocation. 

![arrivalsoc](https://user-images.githubusercontent.com/26678403/26874509-21b6fbf8-4b3b-11e7-8a86-a43b668d46d8.PNG)

**Comment by aesculus** [comment]
> It doesn't make sense that you could leave the Bend supercharger with 33% (according to the Waypoint Inspector) and arrive at the nearby final destination with 32% SoC (that's plausible) and a 19% reserve allocation.

Bug for sure. 32% is the total energy at arrival, of which 19% is reserved for local travel. This gets you to 13% which is too low since you asked for 20% min. Need to look at why this is. It really should be 20% + 19%.

**Comment by aesculus** [comment]
I spoke too soon but there still appears to be a bug.  Here is what I see after looking a bit more:
- Had you not added the Bend SC you would have arrived at Bend from the Detroit Lake SC with 20% + 19%. 
- When you forced the Bend SC into the mix it has a minimum charge time of 15 min. That meant that you were going to get to Bend with 32% arrival SoC base, allowing for 1% loss on the way there. But that is wrong
- I suspect an error in my algorithm that sees that you min charged at the Bend SC and did not do the math correctly adding in the destination requirements.

**Comment by aesculus** [comment]
I fixed this in V1.2 (36). I am not certain why this did not show up before.  We need to keep an eye on this to make sure it stays correct.

![image](https://user-images.githubusercontent.com/283364/26897039-deee26d4-4b7b-11e7-8f6a-919215508cc8.png)

**Comment by EVGrokker** [comment]
I think the challenge was probably the order in which I created the trip. I started with the street address as the destination, then added the supercharger as a waypoint, then added the reserve energy.

**Comment by aesculus** [comment]
Maybe. But the other day I had an issue where I was using an embedded waypoint count within the trip.  Something got screwed up along the way and caused that number to be wrong. So I decided to just switch to counting the waypoints instead figuring the count was equal to the last waypoint order.

This area in the code is suspicious because just above that is the waypoint count thing. I cannot remember if this is the place I changed but none the less.  Having said that the routine to add the destination reserve to the prior waypoint was not right.  It never found it because it was actually trying to apply it to the destination itself.  I have not touched that code in eons so I cannot understand why it worked before unless when I made the new last waypoint order I inadvertently left off the prior -1 which was there before. This is the most likely scenario.
